source/rust_verify/src/singular.rs

```diff
@@ -348,56 +363,81 @@ pub fn check_singular_valid(
-                            assert!(res[0].contains("// ** redefining"));
-                            continue;
-                        } else if res[0].contains("?") {
-                            return ValidityResult::UnexpectedOutput(String::from(format!(
-                                "{} \ngenerated singular query: {}",
-                                res[0].as_str(),
-                                query
-                            )));
-                        } else {
-                            return ValidityResult::Invalid(
-                                None,
-                                err.clone().secondary_label(
-                                    span,
-                                    format!("Singular returned:\n {}", res.join("\n")),
-                                ),
-                            );
-                        }
-                    }
-                }
-            } else {
-                panic!("internal error: integer_ring")
+            if let Err(info) = encoder.encode_requires_poly(expr) {
+                return Err(ValidityResult::Invalid(
+                    None,
+                    err.clone().secondary_label(func_span, info),
+                ));
+            }
+        }
+    }
+
+    // each ensures is a separate query string
+    for stmts in &**enss {
+        if let air::ast::StmtX::Assert(err, expr) = &**stmts {
+            let err: vir::messages::Message =
+                err.clone().downcast().expect("unexpected value in Any -> Message conversion");
+            let res = encoder.encode_ensures_poly(expr);
+            if let Err(info) = res {
+                return Err(ValidityResult::Invalid(
+                    None,
+                    err.clone().secondary_label(func_span, info),
+                ));
             }
+            queries.push((res.unwrap(), err.clone()));
+        }
+    }
+    return Ok(());
+}
+
+pub fn check_singular_valid(
+    context: &mut air::context::Context,
+    command: &Command,
+    func_span: &vir::messages::Span,
+    _query_context: QueryContext<'_, '_>,
+) -> ValidityResult {
+    let mut queries = vec![];
+    if let Err(res) = encode_singular_queries(command, func_span, &mut queries) {
+        // in case of any encoding error, skip running Singular
+        return res;
+    }
+
+    let singular_manager = SingularManager::new();
+    let mut singular_process = singular_manager.launch();
+
+    for (query, err) in queries {
+        air::singular_manager::log_singular(context, &query, &func_span.as_string);
+        let res = singular_process.send_commands(query.as_bytes().to_vec());
+        if (res.len() == 1) && (res[0] == "0") {
+            // this query is ok (poly reduced to 0)
+            continue;
+        } else if (res.len() == 2) && (res[1] == "0") {
+            // this query is switching from the previous one
+            // this will redefine the ring
+            // ignore the first line of the output
```

Comment:
I don't understand what this means: is this a line that appears as a "trailer" to a previous query? Please describe the expected interaction in more detail.

Reply:
I added some more comments on this. Basically, we redefine the ideal we work with, so we expect the redefinition  comment.